Functional test "Stats are being logged for metric" fails

This test randomly fails with:

```
2019-01-24 19:44:00.353 [ERROR]   OpenTsdbSpec.Stats for default rules:40 Condition not satisfied:
2019-01-24 19:44:00.355 
2019-01-24 19:44:00.355 otsdb.query(2.minutes.ago, metric, tags).dps.size() > 0
2019-01-24 19:44:00.356 |     |       |       |    |       |     |   |      |
2019-01-24 19:44:00.356 |     |       |       |    |       |     [:] 0      false
2019-01-24 19:44:00.357 |     |       |       |    |       [cookieHex:0x8000000000000003]
2019-01-24 19:44:00.357 |     |       |       |    pen.switch.flow.system.packets
2019-01-24 19:44:00.358 |     |       |       Thu Jan 24 15:41:59 UTC 2019
2019-01-24 19:44:00.358 |     |       2 minutes
2019-01-24 19:44:00.359 |     StatsResult(metric=pen.switch.flow.system.packets, tags={cookieHex=0x8000000000000003}, aggregateTags=[switchid], dps={})
2019-01-24 19:44:00.360 <org.openkilda.testing.service.otsdb.OtsdbQueryServiceImpl@43ec61f0 restTemplate=org.springframework.web.client.RestTemplate@a3a54d6>
2019-01-24 19:44:00.360 
```

--- a/services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/isl/IslMovementSpec.groovy
+++ b/services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/isl/IslMovementSpec.groovy
@@ -5,10 +5,12 @@ import static org.junit.Assume.assumeTrue
 import static org.openkilda.messaging.info.event.IslChangeType.DISCOVERED
 import static org.openkilda.messaging.info.event.IslChangeType.FAILED
 import static org.openkilda.messaging.info.event.IslChangeType.MOVED
+import static org.openkilda.testing.Constants.STATS_LOGGING_TIMEOUT
 import static org.openkilda.testing.Constants.WAIT_OFFSET
 
 import org.openkilda.functionaltests.BaseSpecification
 import org.openkilda.functionaltests.helpers.Wrappers
+import org.openkilda.testing.Constants.DefaultRule
 import org.openkilda.testing.model.topology.TopologyDefinition
 import org.openkilda.testing.model.topology.TopologyDefinition.Isl
 import org.openkilda.testing.service.lockkeeper.model.ASwitchFlow
@@ -96,6 +98,10 @@ class IslMovementSpec extends BaseSpecification {
         def islToPlugInto = notConnectedIsls[1]
 
         when: "Plug an ISL between two ports on the same switch"
+        def beforeReplugTime = new Date()
+        def dropCounterBefore = northbound.getSwitchRules(islToPlugInto.srcSwitch.dpId).flowEntries.find {
+            it.cookie == DefaultRule.DROP_LOOP_RULE.cookie
+        }.packetCount
         def expectedIsl = islUtils.replug(islToPlug, true, islToPlugInto, true)
 
         then: "The potential self-loop ISL is not present in list of isls (wait for discovery interval)"
@@ -104,6 +110,15 @@ class IslMovementSpec extends BaseSpecification {
         !islUtils.getIslInfo(allLinks, expectedIsl).present
         !islUtils.getIslInfo(allLinks, islUtils.reverseIsl(expectedIsl)).present
 
+        and: "Self-loop rule packet counter is incremented and logged in otsdb"
+        def statsData = null
+        Wrappers.wait(STATS_LOGGING_TIMEOUT, 2) {
+            statsData = otsdb.query(beforeReplugTime, "pen.switch.flow.system.packets",
+                    [switchid: expectedIsl.srcSwitch.dpId.toOtsdFormat(),
+                     cookieHex: DefaultRule.DROP_LOOP_RULE.toHexString()]).dps
+            assert statsData && !statsData.empty
+        }
+        statsData.values().last().toLong() > dropCounterBefore
 
         and: "Unplug the link how it was before"
         lockKeeper.removeFlows([

--- a/services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowPingSpec.groovy
+++ b/services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowPingSpec.groovy
@@ -2,6 +2,7 @@ package org.openkilda.functionaltests.spec.northbound.flows
 
 import static com.shazam.shazamcrest.matcher.Matchers.sameBeanAs
 import static org.junit.Assume.assumeTrue
+import static org.openkilda.testing.Constants.STATS_LOGGING_TIMEOUT
 import static org.openkilda.testing.Constants.WAIT_OFFSET
 import static spock.util.matcher.HamcrestSupport.expect
 
@@ -13,6 +14,7 @@ import org.openkilda.messaging.info.event.SwitchInfoData
 import org.openkilda.northbound.dto.flows.PingInput
 import org.openkilda.northbound.dto.flows.PingOutput.PingOutputBuilder
 import org.openkilda.northbound.dto.flows.UniFlowPingOutput
+import org.openkilda.testing.Constants.DefaultRule
 import org.openkilda.testing.model.topology.TopologyDefinition.Switch
 import org.openkilda.testing.service.lockkeeper.model.ASwitchFlow
 
@@ -35,6 +37,10 @@ class FlowPingSpec extends BaseSpecification {
         flowHelper.addFlow(flow)
 
         when: "Ping the flow"
+        def beforePingTime = new Date()
+        def unicastCounterBefore = northbound.getSwitchRules(srcSwitch.dpId).flowEntries.find {
+            it.cookie == DefaultRule.VERIFICATION_UNICAST_RULE.cookie
+        }.byteCount
         def response = northbound.pingFlow(flow.id, new PingInput())
 
         then: "Ping is successful"
@@ -45,6 +51,16 @@ class FlowPingSpec extends BaseSpecification {
         !response.error
         !response.forward.error
         !response.reverse.error
+
+        and: "Unicast rule packet count is increased and logged to otsdb"
+        def statsData = null
+        Wrappers.wait(STATS_LOGGING_TIMEOUT, 2) {
+            statsData = otsdb.query(beforePingTime, "pen.switch.flow.system.bytes",
+                    [switchid: srcSwitch.dpId.toOtsdFormat(),
+                     cookieHex: DefaultRule.VERIFICATION_UNICAST_RULE.toHexString()]).dps
+            assert statsData && !statsData.empty
+        }
+        statsData.values().last().toLong() > unicastCounterBefore
 
         and: "Remove the flow"
         flowHelper.deleteFlow(flow.id)

--- a/services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/stats/OpenTsdbSpec.groovy
+++ b/services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/stats/OpenTsdbSpec.groovy
@@ -27,25 +27,12 @@ class OpenTsdbSpec extends BaseSpecification {
                  "pen.switch.tx-bytes", "pen.switch.tx-bits", "pen.switch.tx-packets"],
                 uniqueSwitches.collect { [switchid: it.dpId.toOtsdFormat()] }].combinations()
                 + [["pen.isl.latency"], uniqueSwitches.collect { [src_switch: it.dpId.toOtsdFormat()] }].combinations()
-                + [["pen.isl.latency"], uniqueSwitches.collect { [dst_switch: it.dpId.toOtsdFormat()] }].combinations())
+                + [["pen.isl.latency"], uniqueSwitches.collect { [dst_switch: it.dpId.toOtsdFormat()] }].combinations()
+                + [["pen.switch.flow.system.packets", "pen.switch.flow.system.bytes", "pen.switch.flow.system.bits"],
+                   [[cookieHex: DefaultRule.VERIFICATION_BROADCAST_RULE.toHexString()]]].combinations())
     }
 
     def getUniqueSwitches() {
         topology.activeSwitches.unique { it.ofVersion }
-    }
-
-    @Unroll("Stats are being logged for metric:#metric, tags:#tags")
-    def "Stats for default rules"(metric, tags) {
-        expect: "At least 1 result in the past 2 minutes"
-        otsdb.query(2.minutes.ago, metric, tags).dps.size() > 0
-
-        where:
-        [metric, tags] << ([
-                ["pen.switch.flow.system.packets", "pen.switch.flow.system.bytes", "pen.switch.flow.system.bits"],
-                [[cookieHex: DefaultRule.DROP_RULE.toHexString()],
-                 [cookieHex: DefaultRule.VERIFICATION_BROADCAST_RULE.toHexString()],
-                 [cookieHex: DefaultRule.VERIFICATION_UNICAST_RULE.toHexString()],
-                 [cookieHex: DefaultRule.DROP_LOOP_RULE.toHexString()]]
-        ].combinations())
     }
 }

--- a/services/src/test-library/src/main/java/org/openkilda/testing/Constants.java
+++ b/services/src/test-library/src/main/java/org/openkilda/testing/Constants.java
@@ -26,6 +26,7 @@ public final class Constants {
     public static final Integer RULES_INSTALLATION_TIME = 5;
     public static final Integer HEARTBEAT_INTERVAL = 10;
     public static final Integer MAX_DEFAULT_METER_ID = 15;
+    public static final Integer STATS_LOGGING_TIMEOUT = 70;
 
     private Constants() {
         throw new UnsupportedOperationException();

